'Mass file loader' for use in skin and asset loading

ewancg opened the pull request:
<!-- What is the motivation for the changes of this pull request? -->
## Summary:
Currently, asset and skin loading are quite slow if you have a decent amount of them, even on a fast disk. I would like to create a generic, multi-threaded file loader system that can lessen the impact of some necessary disk blocks, and implement it into the skin and asset systems so the brunt of this wait time can be avoided (most or all skins and assets loaded by the time a user finds and joins a server from the browser). This file loader will also be available (and hopefully encouraged) for similar future tasks requiring a lot of files to be loaded & used in a similar way. 

This is a relatively big-picture feature, if my goal is at all unclear to you, I am happy to clarify.

## Goals:
- Flexibility; allowing you to choose which files to or not to load without ever interacting with any `io_*`/`fs_*`/`IStorage::*` - functions.
- Rich error reporting; not only reporting that the operation failed but also why, as well as granting the integrator choice to continue in each instance.
- Simplicity; requiring only a few steps to use, it should be no harder to grasp than a manual/tailored implementation.

At the moment, I have gotten the interface and reference implementation (single-threaded) to a stage where they are ready for review, after which I plan to create the multi-threaded implementation and then add to skin and asset loading.

Just for the sake of it, here is a checklist.
- [x] Interface
- [x] Reference Implementation
- [x] Asynchronous Implementation
- [x] Integration to skin loading
- [ ] Integration to asset loading

## Misc. stuff:
At the moment, the API is described in `src/game/client/file_loader.h`. It is not formatted as proper documentation, I could use some help with that. Maybe it's not needed?

2 functions have been added to `src/base/system.cpp`/`.h` in order to allow for enhanced error checking. It is not strictly needed, but I would like for it to be kept. They have been tested and seen as working on Windows and Linux

There is a simple *temporary* test case in `src/engine/client.cpp`. Up to this point I have been using a test file/directory schema that looks like this
```
.
├── limits
│   ├── 1GiB_file.test.txt 				# 1GiB, expected to succeed in loading
│   ├── 2GiB_file.test.txt 				# 2GiB, expected to fail because it's too big
│   ├── 4GiB_file.test.txt 				# 4GiB, same as above
│   ├── empty_file.test.txt 				# 0 bytes, expected to succeed in loading
│   ├── unreadable_directory 				# expected to fail because it's unreadable
│   └── unreadable_file.test.txt 			# expected to fail because it's unreadable
├── quantity 						# all 250 test files expected to show up
│   ├── 000.test.txt
...
│   └── 250.test.txt
├── recursion_1 					# all recursion tests expected to show up independently
│   ├── recursion_2
│   │   ├── recursion_3
│   │   │   └── status.txt
│   │   └── status.txt
│   └── status.txt
└── status.txt
```

I am trying my best to stick to DDNet code convention. Please let me know if there's something you think doesn't fit.

<!-- Note that builds and other checks will be run for your change. Don't feel intimidated by failures in some of the checks. If you can't resolve them yourself, experienced devs can also resolve them before merging your pull request. -->

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

ewancg commented:
I would like to know if it's wanted in the DDNet code & if the way I'm doing it is considered to be the 'right approach' before I go on making the multi-threaded implementation & replacing the asset and skin loading systems with it. I will wait for feedback before making major changes.

def- commented:
Would it make sense to use aio for that?

ewancg commented:
I don't know the extent to which aio has been implemented, but I hear it doesn't work on Windows. If it only works file-by-file, this still has the benefit of being batch-oriented

Chairn commented on src/base/system.cpp:
```diff
@@ -2360,6 +2364,98 @@ int fs_is_relative_path(const char *path)
 #endif
 }
 
+int fs_is_readable(const char *path)
```
isn't it easier to just try opening the file?

ewancg replied:
this (a.) also applies to directories, and (b., in the instance where I use it) can be caught while roughing out which files are & aren't eligible, which is a different step from actually reading the files. not completely objected to opening it twice, since that's what this ultimately does, but i would like to think that this approach is 'cleaner'

also it can easily be expanded to test for write and exec perms if need be, not too worried about the presence of this func

Robyt3 replied:
I'd expect that checking before every file operation brings a significant performance overhead when there are many small files like skins. I don't think checking before opening a file/folder is very useful. You can still try to find out more details after the error already occured instead of checking for potential errors first.

heinrich5991 replied:
> consider this (a.) also applies to directories, and (b., in the instance where I use it) can be caught while roughing out which files are & aren't eligible, which is a different step from actually reading the files. not completely objected to opening it twice, since that's what this ultimately does, but i would like to think that this approach is 'cleaner'

Why is that a separate step, actually?

ewancg replied:
It currently sifts through the provided paths and adds all eligible files to an index, because it seemed simpler that way. I don't think I'm going to change this, but I will make sure that there's only one file open/read and I will just error at the time of actually opening

Robyt3 commented on src/base/system.cpp:
```diff
@@ -2360,6 +2364,98 @@ int fs_is_relative_path(const char *path)
+	DWORD generic_access_rights = GENERIC_READ;
+	DWORD security_descriptor_size = 0;
+
+	if(!GetFileSecurity(wide_path.c_str(), OWNER_SECURITY_INFORMATION | GROUP_SECURITY_INFORMATION | DACL_SECURITY_INFORMATION, NULL, 0, &security_descriptor_size) &&
```
Always use the `W` functions from the Windows API explicitly.

`GetFileSecurity` -> `GetFileSecurityW` (also below)

ewancg replied:
Removed in this commit

heinrich5991 commented:
Sounds like a good idea in general.

High-level: I think stuff like "is_readable" shouldn't be used, rather, the file can be opened in the background thread and the resulting error can be reported. `std::regex` is known to be slow and should probably not be used.

Is there a reason for denying symlinks by default or at all?

ewancg commented:
Recursing through symlinks presents the risk of looping forever.

ewancg commented:
I've compounded it into all step & removed `fs_is_readable`. I'm now unable to check for unreadable directories at all and unreadable files if the flags contain `LOAD_FLAGS_DONT_READ_FILE`.

Kaffeine commented:
>> heinrich5991: Is there a reason for denying symlinks by default or at all?
> ewancg: Recursing through symlinks presents the risk of looping forever.

ewancg commented:
I planned to have them on for skins. What's the big deal

Kaffeine commented:
I just don't see the need in implementing the loader and special symlinks processing in one MR.

> Recursing through symlinks presents the risk of looping forever.

It would be a 100% user fault, and it does not look like a new problem. Why we'd mix it into a new feature.

Robyt3 commented on src/game/client/file_loader.h:
```diff
@@ -0,0 +1,236 @@
+		// If continued, the error is ignored but is likely to happen again.
+		// pUser = nullptr
+
+		LOAD_ERROR_NOT_INIT,
+		// Implementation-specific load function called with any combination of the following issues: no file load callback, no paths, invalid IStorage pointer, or invalid flags.
+		// This is the only error where the return value is inconsequential.
+		// pUser = nullptr
```
I think all of these cases should instead be checked with `dbg_assert` and immediately crash, because they are signs of programming errors that should be detected early.

ewancg replied:
Definitely not all of them. Some of them cannot be detected before they error so there should at least be some opportunity for recovery. The only one I can think of at the moment I agree with turning into an assert is a failed load since it's already unrecoverable. Things like the invalid path error can depend on whether a directory exists, which can be influenced by a lot more than just dev error; imagine it's a part of a bigger operation, it would be believable that the path doesn't exist (batch file download? archive extraction? could be any of many possibilities)

ewancg replied:
Sorry. I'm on my phone and misread the block your review pertains to. Sure, I'll change it

ewancg commented:
I'll start the async implementation soon if there are no more complaints.

Robyt3 commented on src/game/client/file_loader.cpp:
```diff
@@ -65,15 +65,11 @@ inline bool CompareExtension(const std::filesystem::path &Filename, const std::s
-		TryCallback<bool>(m_fnLoadFailedCallback, LOAD_ERROR_NOT_INIT, nullptr);
-		return 0;
-	}
+	dbg_assert(!m_RequestedPaths.empty() /* Paths have been added */
+			   && m_pStorage /* Storage is valid */
+			   && m_fnFileLoadedCallback /* File loaded callback is implemented */
+			   && !(m_Flags ^ LOAD_FLAGS_MASK) /* Flags are in bounds */,
+		"Mass file loader used without proper initialization.");
```
Turn each into an individual `dbg_assert` so it's more clear what specifically went wrong.

heinrich5991 commented:
Status: Draft.

Robyt3 commented on src/game/client/file_loader.h:
```diff
@@ -0,0 +1,225 @@
+#ifndef GAME_CLIENT_FILE_LOADER_H
```
This doesn't seem to do anything game related, so it should be moved to `src/engine/shared` so it could also be used in the server or tools.

ewancg replied:
moved 😄

Robyt3 commented on src/base/system.cpp:
```diff
@@ -71,16 +72,19 @@
 #include <winsock2.h>
 #include <ws2tcpip.h>
 
+#include <aclapi.h> // fs_is_readable
 #include <cerrno>
 #include <cfenv>
 #include <io.h>
 #include <objbase.h>
 #include <process.h>
+#include <securitybaseapi.h> // fs_is_readable
 #include <share.h>
 #include <shellapi.h>
 #include <shlobj.h> // SHChangeNotify
 #include <shlwapi.h>
 #include <wincrypt.h>
+#include <winioctl.h> // fs_is_symlink
```
I think the comments on the includes are more useful if they show which external function/struct/etc. they were required for (see `#include <shlobj.h> // SHChangeNotify`).

ewancg replied:
added which constant the include was needed for. thanks for bringing this up, i forgot i didn't need aclapi.h or securitybaseapi.h anymore

ewancg commented:
I've added the basic async implementation.
Here are some things that I am really unhappy about
- Mysterious SIGILL from HalRequestSoftwareInterrupt shortly after entering the game; I am probably forgetting to access something thread-safely, and it's causing bad things to happen. I haven't had time to diagnose this yet
- File loader thread has to wait some ms for the main thread to get a chance to access the textures queued for upload to the GPU, otherwise both threads just take turns hogging the mutex, making the game unplayable. It is also not ideal to have a hard-baked time constant, and I think this is generally the wrong approach; I want to find a lockfree way to do this.

I have added locks to outside CSkins uses as needed. I also commented out some GUI warnings in the graphics code in favor of a console warning. If it's decided that these UI warnings are still needed, I am happy to add an option (I didn't want to go poking around adding an optional warn param. to the relevant functions because that might not be the right approach, and that would require changing the interface def.)

ewancg commented:
I am willing to continue working on this as soon as I can get some more insight into whether or not I am doing this the desired way. Any help would be appreciated. We are nearing the point where if I don't start working on this again soon I'm gonna forget all about it.

Robyt3 commented on src/engine/shared/file_loader.cpp:
```diff
@@ -0,0 +1,194 @@
+					continue;
+				}
+
+				// system.cpp APIs are only good up to 2 GiB at the moment (signed long cannot describe a size any larger)
+				io_seek(Handle, 0, IOSEEK_END);
+				long ExpectedSize = io_tell(Handle);
+				io_seek(Handle, 0, IOSEEK_START);
```
Seems redundant and like a TOCTOU bug. It's better to just read the entire file immediately with `IStorage::ReadFile`.

Robyt3 replied:
I'm not sure whether it's a good idea to always load the files into memory though. Some of our loaders only work with files, like datafiles (maps).

ewancg replied:
I don't think this is a TOCTOU bug. This is a very common way to probe file size (and as far as I know, the only way using the libc like the system APIs here are doing). Unless you mean I shouldn't be checking the file size at all?

Robyt3 replied:
Yeah, there is no reason to check the size first. The `io_read_all` and `ReadFile` functions ensure that all that can be read is read.

ewancg replied:
I think I originally added this so I could consistently differentiate between empty files (where the size probe gives 0) and files that are too big to be opened (where the bytes read reports as 0 on some compilers)

Robyt3 replied:
Does `io_length` return `-1` for empty files? If so, that's a bug and should be fixed separately. If not, then only use `io_read_all` and remove `ExpectedSize`.

If you want to check for too large files (which is unnecessary in my opinion), then maybe you could check in `io_read_all` instead (in a separate PR).

ewancg replied:
It returns -1 for files larger than 2GB on some compilers.

heinrich5991 replied:
I see no reason for all this code tbh. It seems to just call `io_read_all`. If `io_read_all` has a bug, then that bug should be fixed, and not worked around.

Robyt3 commented on src/engine/shared/file_loader.cpp:
```diff
@@ -0,0 +1,194 @@
+			if(m_Continue)
+			{
+				std::string FilePath = Directory.first + "/" + File;
+				if(!(m_Flags & LOAD_FLAGS_FOLLOW_SYMBOLIC_LINKS) && fs_is_symlink(FilePath.c_str()))
+				{
+					m_Continue = TryCallback(m_fnLoadFailedCallback, LOAD_ERROR_UNWANTED_SYMLINK, FilePath.c_str(), m_pUser);
+					continue;
+				}
```
Why do we want to conditionally exclude symlinks?

ewancg replied:
Possible infinite recursion

Kaffeine replied:
It was not a (n our) problem before — user can break the files in any way and we're not responsible for that.
I actively use symlinks for valid reasons and don't see why we need to care about it.

ewancg replied:
I think I will just remove symlink detection

Robyt3 commented on src/engine/shared/file_loader.cpp:
```diff
@@ -0,0 +1,194 @@
+				// 	continue;
+				// }
+
+				Handle = io_open(FilePath.c_str(), IOFLAG_READ | IOFLAG_SKIP_BOM);
```
You'd need another flag to decide whether to use `ReadFile` or `ReadFileStr` because we don't want to skip the BOM on all files.

Robyt3 commented on src/engine/shared/file_loader.cpp:
```diff
@@ -0,0 +1,194 @@
+	}
+}
+
+void CMassFileLoader::SetFileExtension(const std::string &Extension)
```
Don't use `std::string` unless there is a very good reason. The only warranted use of `std::string` I see is `std::set<std::string>`. For protoyping you can keep it though, since it might be easier to write.

ewancg replied:
What about a string view here?

Robyt3 commented on src/engine/shared/file_loader.cpp:
```diff
@@ -0,0 +1,194 @@
+
+[[maybe_unused]] int CMassFileLoader::ListDirectoryCallback(const char *Name, int IsDir, int, void *User)
+{
+	auto *UserData = reinterpret_cast<SListDirectoryCallbackUserInfo *>(User);
```
`pUserData` (the naming convention also applies to auto variables)

Robyt3 commented on src/engine/shared/file_loader.cpp:
```diff
@@ -0,0 +1,194 @@
+			UserData->m_pThis->m_PathCollection.insert({AbsolutePath, new std::vector<std::string>});
+			// Note that adding data to a SORTED container that is currently being iterated on higher in scope would invalidate the iterator. This is not sorted
+			SListDirectoryCallbackUserInfo Data{&AbsolutePath, UserData->m_pThis, UserData->m_pContinue};
+			UserData->m_pThis->m_pStorage->ListDirectory(IStorage::TYPE_ALL, AbsolutePath.c_str(), ListDirectoryCallback, &Data); // Directory item is a directory, must be recursed
```
The storage type has to be the one used initially instead of `TYPE_ALL` every time.

ewancg replied:
Not sure what you mean. It's TYPE_ALL initially as well, in the direct ListDirectory invocation in the load function

Robyt3 replied:
What is mean is that the loader should have a parameter `int StorageType` so the storage type that is listed can be adjusted, same as the existing `ListDirectory`.

ewancg replied:
That's why the ":" exists on a per-directory basis. That is the solution I cane up with and thought everyone was happy with when I realized it would be tricky to provide the storage type for each path.

Robyt3 commented on src/engine/shared/file_loader_async.cpp:
```diff
@@ -0,0 +1,196 @@
+
+	char aAuxBuf[512];
+	str_format(aAuxBuf, sizeof(aAuxBuf), "Unable to create file loader thread with id \"%s\"", aId);
+	dbg_assert(thread_init_and_detach(CMassFileLoaderAsync::Load, this, aId) != 0, aAuxBuf);
```
This should probably use the `IJob` system instead of raw threads.

ewancg commented:
> File loader thread has to wait some ms for the main thread to get a chance to access the textures queued for upload to the GPU, otherwise both threads just take turns hogging the mutex, making the game unplayable. It is also not ideal to have a hard-baked time constant, and I think this is generally the wrong approach; I want to find a lockfree way to do this.

Advice?

ewancg commented:
I can see how regex could be considered overkill, but maybe I should add the ability to filter by more than one extension. Sort of like how most file dialog APIs work. Thoughts?

Robyt3 commented:
> I can see how regex could be considered overkill, but maybe I should add the ability to filter by more than one extension. Sort of like how most file dialog APIs work. If so, multiple strings or one string separated by semicolons? Thoughts?

Not sure it's useful anywhere to list files of multiple extensions at the same time. Unless we can identify a clear use case, I think it's better to keep it simple with only one file extension being filtered.

ewancg commented:
Async implementation pretty much finished. Changed the mutex tug-of-war to a mostly lockfree system else using an `IJob`. Specifically, instead of both threads waiting to access a list of textures (main thread to read; file loader thread to write), the file loader thread goes into 'yield mode' after each file is loaded & texture is prepared, where the main thread can then do what it has to do & set the file loader thread's status back to running.

Waiting for review, I think that this is the final iteration of the file loader itself so that would probably be the best thing to look at. Skins loading also works and could use some review, not 100% sure if that's how it'll end up looking but it works.

Still have to do the assets loading.

Robyt3 commented on src/engine/shared/file_loader.h:
```diff
@@ -0,0 +1,311 @@
+
+	std::vector<std::string> m_RequestedPaths;
+	std::unordered_map<std::string, std::vector<std::string> *> m_PathCollection;
+	char *m_pExtension = nullptr;
```
`m_aExtension[IO_MAX_PATH_LENGTH]`

There's no need for dynamic memory allocation for this. We know the extension can only have a certain maximum size.

heinrich5991 commented:
Status: Not done, waiting for author.

ewancg commented:
Doesn't seem like a high demand feature. Especially considering how much work it is just to get it working. I will consider a simpler/less versatile approach for async skins and assets in the future. Closing for now
